gear/csv_to_json_transformer/src/python/csv_app/transformer.py

```diff
@@ -0,0 +1,238 @@
+        """
+        self.__project = project
+        self.__error_writer = error_writer
+        self.__pending_visits: Dict[str, VisitMapping] = {}
+        self.system_errors = False
+
+    def _update_file_metadata(self, file: FileEntry,
+                              input_record: Dict[str, Any]) -> bool:
+        """Set file modality and info.forms.json metadata.
+
+        Args:
+            file: Flywheel file object
+            input_record: input visit data
+        """
+
+        # remove empty fields
+        non_empty_fields = {
+            k: v
+            for k, v in input_record.items() if v is not None
+        }
+        info = {"forms": {"json": non_empty_fields}}
+        try:
+            file.update(modality='Form')
+            file.update_info(info)
+        except ApiException as error:
+            log.error('Error in setting file %s metadata - %s', file.name,
+                      error)
+            return False
+
+        return True
+
+    def _is_duplicate_record(self, input_record: Dict[str, Any],
+                             existing_file: FileEntry) -> bool:
+        """Check whether the input data matches with an existing visit file.
+
+        Args:
+            input_record: input visit data
+            existing_file: existing visit file
+
+        Returns:
+            True if a duplicate detected, else false
+        """
+        input_dict = sorted(input_record.items())
+        try:
+            currnt_dict = sorted(json.loads(existing_file.read()).items())
+            return (input_dict == currnt_dict)
+        except (json.JSONDecodeError, ApiException) as error:
+            log.warning('Error in reading existing file - %s', error)
+            return False
+
+    def _add_pending_visit(self, *, subject: SubjectAdaptor, filename: str,
+                           file_id: str, input_record: Dict[str, Any]):
+        """Add the visit to the list of visits pending for QC for the
+        participant.
+
+        Args:
+            subject: Flywheel subject adaptor for the participant
+            filename: Flywheel aquisition file name
+            file_id: Flywheel aquisition file ID
+            input_record: input visit data
+        """
+        visit_mapping: VisitMapping
+        subject_lbl = input_record[FieldNames.NACCID]
+        if subject_lbl in self.__pending_visits:
+            visit_mapping = self.__pending_visits[subject_lbl]
+            visit_mapping['visits'].add_visit(
+                filename=filename,
+                file_id=file_id,
+                visitdate=input_record[FieldNames.DATE_COLUMN])
+        else:
+            participant_visits = ParticipantVisits.create_from_visit_data(
+                filename=filename, file_id=file_id, input_record=input_record)
+            visit_mapping = {'subject': subject, 'visits': participant_visits}
+            self.__pending_visits[subject_lbl] = visit_mapping
+
+    def transform_record(self, input_record: Dict[str, Any],
```

Comment:
I would like to separate the transformation from the file upload. Or at a minimum change the name of this method to imply we are creating the JSON file.

Also, we should factor out common code with EnrollmentSubject.upload_enrollment and create a method in SubjectAdaptor